woocommerce-gateway-stripe.php

```diff
@@ -593,16 +599,44 @@ public function register_routes() {
 					$upe_flag_toggle_controller = new WC_Stripe_REST_UPE_Flag_Toggle_Controller();
 					$upe_flag_toggle_controller->register_routes();
 
-					$gateway = WC()->payment_gateways()->payment_gateways()[ WC_Gateway_Stripe::ID ];
-
 					require_once WC_STRIPE_PLUGIN_PATH . '/includes/admin/class-wc-rest-stripe-settings-controller.php';
-					$settings_controller = new WC_REST_Stripe_Settings_Controller( $gateway );
+					$settings_controller = new WC_REST_Stripe_Settings_Controller( $this->get_main_stripe_gateway() );
 					$settings_controller->register_routes();
 
 					$stripe_account_keys_controller = new WC_REST_Stripe_Account_keys_Controller();
 					$stripe_account_keys_controller->register_routes();
 				}
 			}
+
+			/**
+			 * Returns the main Stripe payment gateway class instance.
+			 *
+			 * @return WC_Stripe_Payment_Gateway
+			 */
+			public function get_main_stripe_gateway() {
+				if ( ! is_null( $this->stripe_gateway ) ) {
+					return $this->stripe_gateway;
+				}
+
+				if ( WC_Stripe_Feature_Flags::is_upe_preview_enabled() && WC_Stripe_Feature_Flags::is_upe_checkout_enabled() ) {
+					$this->stripe_gateway = new WC_Stripe_UPE_Payment_Gateway();
+
+					return $this->stripe_gateway;
+				}
+
+				$this->stripe_gateway = new WC_Gateway_Stripe();
+
+				return $this->stripe_gateway;
+			}
+
+			/**
+			 * Use only in tests.
+			 * Running this will not unregister the hooks set by the gateway class.
+			 */
+			// phpcs:ignore WordPress.NamingConventions.ValidFunctionName.MethodNameInvalid
+			public function _TESTS_ONLY__unset_main_stripe_gateway() {
```

Comment:
I needed this so that tests wouldn't fail in `tests/phpunit/test-wc-stripe.php`

Reply:
I think there's a better hack we could use with https://www.php.net/manual/en/closure.bind.php I'll experiment.